In [28]:
import pandas as pd
import numpy as np
import os
import unicodedata

In [29]:
REPO_URL = 'https://github.com/LuisBetancourt11/Proyecto_Ciencia_Datos'
REPO_DIR = 'Proyecto_Ciencia_Datos'

if not os.path.isdir(REPO_DIR):
    !git clone {REPO_URL}
else:
    print("El repositorio ya está clonado")
    !cd {REPO_DIR} && git pull

El repositorio ya está clonado
Already up to date.


In [30]:
raw_dir = REPO_DIR + '/data/raw/'
processed_dir = REPO_DIR + '/data/processed/'

# Asegurar que la carpeta de salida exista
os.makedirs(processed_dir, exist_ok=True)

print("raw_dir      :", raw_dir)
print("processed_dir:", processed_dir)
print("¿Existe el CSV?:", os.path.exists(raw_dir + 'matriculas.csv'))

raw_dir      : Proyecto_Ciencia_Datos/data/raw/
processed_dir: Proyecto_Ciencia_Datos/data/processed/
¿Existe el CSV?: True


In [31]:
df = pd.read_csv(raw_dir + 'matriculas.csv')
print(df.shape)
df.head()

(13461, 21)


,Género,Municipio de Nacimiento,País,Municipio Dirección,Barrio Dirección,Estrato,Estado civil,EPS,Grupo Sisben,Zona,...,Ocupación,Medio de Transporte,Multiculturalidad,Estado,Fecha de Matrícula,Jornada,Programa,Período,Nivel,Edad
0,Femenino,Medellín,Colombia,Envigado,Las Flores,3.0,Soltero(a),EPS Sura,Sin respuesta,Urbana,...,Estudiante básica,A Pie,No aplica,Graduado,2018-02-05T10:00:00Z,MEDIA TÉCNICA,MEDIA TÉCNICA TECNICO LABORAL AUXILIAR ADMINIS...,2018-2,SEMESTRE 2,20.0
1,Femenino,Envigado,Colombia,Envigado,San José,2.0,Soltero(a),EPS Sura,Sin respuesta,Urbana,...,Estudiante superior,Público Metro,No aplica,Graduado,2016-01-19T10:00:00Z,NOCHE,10T-TÉCNICO LABORAL EN AUXILIAR ADMINISTRATIVO...,2016-2,SEMESTRE 2,29.0
2,Masculino,Bogotá,Colombia,La Mesa,Sin respuesta,3.0,Soltero(a),No Aplica,Sin respuesta,Rural,...,Estudiante superior,Transmilenio,No aplica,Cancelado,2021-02-08T10:00:00Z,NOCHE,09T-TÉCNICO LABORAL EN AUXILIAR ADMINISTRATIVO,2021-2,SEMESTRE 1,19.0
3,Masculino,Envigado,Colombia,Envigado,San José,2.0,Soltero(a),SISBEN,B3,Urbana,...,Estudiante superior,Público Bus o Buseta,No aplica,Graduado,2020-01-17T10:00:00Z,DIURNA,01T-TÉCNICO LABORAL AUXILIAR DE TALENTO HUMANO,2021-1,SEMESTRE 2,21.0
4,Masculino,Medellín,Colombia,Envigado,El Salado,2.0,Soltero(a),Nueva Promotora de Salud - Nueva EPS,N0,Urbana,...,Estudiante superior,Moto,No aplica,Inactivo,2020-01-27T10:00:00Z,NOCHE,27T-TÉCNICO LABORAL EN MERCADEO Y VENTAS (Estu...,2020-1,SEMESTRE 1,20.0


In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13461 entries, 0 to 13460
Data columns (total 21 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Género                   13461 non-null  object 
 1   Municipio de Nacimiento  13461 non-null  object 
 2   País                     13461 non-null  object 
 3   Municipio Dirección      13407 non-null  object 
 4   Barrio Dirección         13461 non-null  object 
 5   Estrato                  12193 non-null  float64
 6   Estado civil             13461 non-null  object 
 7   EPS                      13460 non-null  object 
 8   Grupo Sisben             13461 non-null  object 
 9   Zona                     12160 non-null  object 
 10  Nivel de formación       13461 non-null  object 
 11  Ocupación                11543 non-null  object 
 12  Medio de Transporte      13461 non-null  object 
 13  Multiculturalidad        13461 non-null  object 
 14  Estado                

## Estandarización de nombres de columnas

Las columnas vienen con tildes, mayúsculas y espacios (`Municipio de Nacimiento`,
`Fecha de Matrícula`, ...). Las normalizamos a `snake_case` sin tildes para trabajar
más cómodamente y evitar errores de tipeo. Se hace de forma programática (no a mano)
porque son 21 columnas.


In [33]:
def normalizar(texto):
    """minúsculas, sin tildes, espacios -> guion bajo"""
    texto = texto.strip().lower()
    # quitar tildes / acentos
    texto = unicodedata.normalize('NFKD', texto).encode('ascii', 'ignore').decode('ascii')
    texto = texto.replace(' ', '_')
    return texto

df.columns = [normalizar(c) for c in df.columns]
df.columns.tolist()

['genero',
 'municipio_de_nacimiento',
 'pais',
 'municipio_direccion',
 'barrio_direccion',
 'estrato',
 'estado_civil',
 'eps',
 'grupo_sisben',
 'zona',
 'nivel_de_formacion',
 'ocupacion',
 'medio_de_transporte',
 'multiculturalidad',
 'estado',
 'fecha_de_matricula',
 'jornada',
 'programa',
 'periodo',
 'nivel',
 'edad']

## Tratamiento de nulos "disfrazados"

En este dataset los faltantes no vienen como `NaN`, sino como texto:
`Sin respuesta`, `No Aplica`, `No aplica`, etc. Si no los convertimos a nulos reales,
`df.isnull()` no los detecta y contaminan el análisis.

In [34]:
df.isnull().sum()

,0
genero,0
municipio_de_nacimiento,0
pais,0
municipio_direccion,54
barrio_direccion,0
estrato,1268
estado_civil,0
eps,1
grupo_sisben,0
zona,1301


In [35]:
# Cadenas que representan ausencia de dato
sentinelas = ['sin respuesta', 'sin informacion', 'sin información',
              'no aplica', 'no aplica.', 'na', 'n/a', 'sin dato', 'ninguno']

obj_cols = df.select_dtypes(include='object').columns
for col in obj_cols:
    mask = df[col].astype(str).str.strip().str.lower().isin(sentinelas)
    df.loc[mask, col] = np.nan

print("Nulos reales tras normalizar los textos faltantes:")
df.isnull().sum().sort_values(ascending=False)

Nulos reales tras normalizar los textos faltantes:


,0
multiculturalidad,12628
grupo_sisben,10347
nivel_de_formacion,3398
medio_de_transporte,3363
eps,3089
ocupacion,1965
estado_civil,1495
barrio_direccion,1372
zona,1327
estrato,1268


## Estandarización de tipos de datos

- `estrato` y `edad` vienen como `float` (3.0, 20.0). Los pasamos a entero anulable
  (`Int64`), que admite nulos.
- `fecha_de_matricula` viene como texto ISO (`2018-02-05T10:00:00Z`). La parseamos a
  datetime.

In [36]:
for col in ['estrato', 'edad']:
    if col in df.columns:
        df[col] = pd.to_numeric(df[col], errors='coerce').astype('Int64')

df['fecha_de_matricula'] = pd.to_datetime(df['fecha_de_matricula'], errors='coerce',
                                          utc=True).dt.tz_localize(None)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13461 entries, 0 to 13460
Data columns (total 21 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   genero                   13461 non-null  object        
 1   municipio_de_nacimiento  12305 non-null  object        
 2   pais                     13409 non-null  object        
 3   municipio_direccion      13407 non-null  object        
 4   barrio_direccion         12089 non-null  object        
 5   estrato                  12193 non-null  Int64         
 6   estado_civil             11966 non-null  object        
 7   eps                      10372 non-null  object        
 8   grupo_sisben             3114 non-null   object        
 9   zona                     12134 non-null  object        
 10  nivel_de_formacion       10063 non-null  object        
 11  ocupacion                11496 non-null  object        
 12  medio_de_transporte      10098 n

## Normalización de la variable `periodo`

La columna `periodo` viene con tres problemas de formato:

| Problema | Ejemplos | Registros |
|---|---|---|
| Espacios sobrantes | `2015- 1`, `2016 - MT` | ~1,900 |
| Etiqueta `MT` / `MEDIA TÉCNICA` en vez del semestre | `2018-MT`, `2015- MEDIA TÉCNICA` | 2,472 |
| Formato correcto | `2019-1`, `2020-2` | el resto |

**Criterio de normalización.** El sufijo `MT` identifica a los estudiantes de Media Técnica, pero
no indica en qué semestre se matricularon, así que rompe la comparabilidad temporal de la variable.
Como sí tenemos la `fecha_de_matricula`, deducimos el semestre a partir del mes:

- Meses **1 a 6** (enero-junio) → semestre **1**
- Meses **7 a 12** (julio-diciembre) → semestre **2**

Ejemplo: un registro `2018-MT` con fecha de matrícula en enero pasa a ser `2018-1`.

**¿Se pierde la información de Media Técnica?** No. Se verificó que 2,457 de los 2,472 registros con
periodo `MT` ya tienen `jornada = MEDIA TÉCNICA`, así que esa condición sigue identificable por la
columna `jornada` y por `programa`. Adicionalmente se conserva la bandera `periodo_era_mt`.

**Caso sin solución:** 168 registros con periodo `2026-MT` no tienen `fecha_de_matricula`, por lo que
la regla no puede aplicarse. Se dejan marcados y se documentan como limitación.

In [37]:
print("Periodos ANTES de normalizar:", df['periodo'].nunique(), "valores únicos")
print(df['periodo'].value_counts(dropna=False).sort_index().to_string())

# --- 1) Limpiar espacios sobrantes: '2015- 1' -> '2015-1', '2016 - MT' -> '2016-MT' ---
df['periodo'] = df['periodo'].astype('string').str.replace(r'\s+', '', regex=True)

# --- 2) Unificar la etiqueta de Media Técnica ('MEDIA TÉCNICA' -> 'MT') ---
df['periodo'] = df['periodo'].str.replace('MEDIATÉCNICA', 'MT', regex=False)
df['periodo'] = df['periodo'].str.replace('MEDIATECNICA', 'MT', regex=False)

# --- 3) Marcar los registros MT antes de convertirlos (trazabilidad) ---
df['periodo_era_mt'] = df['periodo'].str.contains('MT', na=False).astype(int)
n_mt = df['periodo_era_mt'].sum()

# --- 4) Convertir MT -> semestre deducido del mes de matrícula ---
es_mt = df['periodo_era_mt'] == 1
tiene_fecha = df['fecha_de_matricula'].notna()

anio_periodo = df['periodo'].str.extract(r'^(\d{4})')[0]          # año que ya trae el periodo
mes = df['fecha_de_matricula'].dt.month
semestre = np.where(mes <= 6, '1', '2')                            # 1-6 -> sem 1 | 7-12 -> sem 2

convertibles = es_mt & tiene_fecha
df.loc[convertibles, 'periodo'] = (anio_periodo[convertibles] + '-' + pd.Series(semestre, index=df.index)[convertibles])

n_convertidos = convertibles.sum()
n_sin_fecha = (es_mt & ~tiene_fecha).sum()

print()
print("═" * 58)
print("NORMALIZACIÓN DE PERIODO")
print("═" * 58)
print(f"Registros con etiqueta MT:            {n_mt}")
print(f"  Convertidos con la regla del mes:   {n_convertidos}")
print(f"  Sin fecha (no convertibles):        {n_sin_fecha}")
print()
print("Periodos DESPUÉS de normalizar:", df['periodo'].nunique(), "valores únicos")
print(df['periodo'].value_counts(dropna=False).sort_index().to_string())

# --- 5) Variables derivadas listas para el análisis ---
df['anio_periodo'] = pd.to_numeric(df['periodo'].str.extract(r'^(\d{4})')[0], errors='coerce').astype('Int64')
df['semestre_periodo'] = pd.to_numeric(df['periodo'].str.extract(r'-(\d)$')[0], errors='coerce').astype('Int64')

print()
print("Verificación — reparto por semestre tras la conversión:")
print(df['semestre_periodo'].value_counts(dropna=False).sort_index().to_string())

Periodos ANTES de normalizar: 25 valores únicos
periodo
2014-2                  79
2015- 1                410
2015- MEDIA TÉCNICA      6
2015-2                 620
2016 - MT              104
2016-1                 573
2016-2                 681
2017 - MT               68
2017-1                 763
2017-2                 805
2018 - MT              272
2018-1                 526
2018-2                 708
2019 - MT               89
2019-1                 689
2019-2                 600
2020 - MT              930
2020-1                 534
2020-2                 905
2021 - MT              857
2021-1                 625
2021-2                 583
2022-1                 817
2026-1                 682
2026-MT                169
NaN                    366

══════════════════════════════════════════════════════════
NORMALIZACIÓN DE PERIODO
══════════════════════════════════════════════════════════
Registros con etiqueta MT:            2495
  Convertidos con la regla del mes:   2326
  Sin fecha 

# **Limpieza de datos invalidos y normalizar categorías**
## **El estrato: imputación jerárquica por barrio**
En Colombia los estratos válidos van de 1 a 6, el valor 0 es un error de registro (46 casos)
y además hay 1,268 registros sin estrato. Diagnóstico: esos nulos no son aleatorios el 90%
de los cursos de extensión no tiene estrato (no se recogió el formulario socioeconómico), mientras
que en técnicos laborales el faltante es solo 1.2%.

Criterio de imputación. En Colombia el estrato se asigna por la ubicación geográfica del predio,
por lo que el barrio de residencia es el mejor predictor disponible del estrato. En vez de usar
una mediana global (que asigna el mismo valor a todos) aplicamos una cascada de tres niveles:

| Nivel | Criterio | Cuándo aplica |
|---|---|---|
| 1 | Mediana del barrio (mínimo 10 registros con estrato conocido) | El estudiante tiene barrio identificable |
| 2 | Mediana del municipio de residencia | Barrio faltante o barrio sin muestra suficiente |
| 3 | Mediana global | No hay ni barrio ni municipio |

La tabla de referencia barrio→estrato se construye con los propios datos (11,253 registros que
sí traen barrio y estrato), lo que evita depender de una fuente externa y refleja la población real
de la institución. La consistencia interna es alta: en barrios como San Rafael el 90% de los
registros comparten el mismo estrato, en Barrio Mesa el 88%.

Se conserva la bandera estrato_imputado (1 = valor imputado) para no perder la señal de que el
dato no estaba registrado.

## **Medio de transporte**
La columna viene con 20 categorías duplicadas, que se agrupan en 8 estándar para reducir ruido y
facilitar la codificación del modelo. Los nulos se imputan con la moda condicional por grupo.

In [38]:
# ═══════════════════════════════════════════════════════════════
# 1) ESTRATO — Imputación jerárquica (barrio → municipio → global)
# ═══════════════════════════════════════════════════════════════
MIN_N = 10  # mínimo de registros para que un grupo sea confiable como referencia

n_estrato_0  = (df['estrato'] == 0).sum()
n_estrato_na = df['estrato'].isna().sum()

# Bandera: marca los registros que serán imputados (antes de tocar nada)
df['estrato_imputado'] = (df['estrato'].isna() | (df['estrato'] == 0)).astype(int)

# El 0 es inválido en Colombia -> se trata como faltante para no contaminar las medianas
df['estrato'] = df['estrato'].replace(0, np.nan)

# --- NIVEL 1: mediana del barrio ---
tabla_barrio = (df[df['estrato'].notna()]
                .groupby('barrio_direccion')['estrato']
                .agg(['median', 'count'])
                .query(f'count >= {MIN_N}')['median'].round())  # redondeo: la mediana puede dar 2.5

estimado_barrio = df['barrio_direccion'].map(tabla_barrio)
n_nivel1 = (df['estrato'].isna() & estimado_barrio.notna()).sum()
df['estrato'] = df['estrato'].fillna(estimado_barrio)

# --- NIVEL 2: mediana del municipio (solo con datos reales, no imputados) ---
reales = df['estrato_imputado'] == 0
tabla_municipio = (df[reales & df['estrato'].notna()]
                   .groupby('municipio_direccion')['estrato']
                   .agg(['median', 'count'])
                   .query(f'count >= {MIN_N}')['median'].round())

estimado_municipio = df['municipio_direccion'].map(tabla_municipio)
n_nivel2 = (df['estrato'].isna() & estimado_municipio.notna()).sum()
df['estrato'] = df['estrato'].fillna(estimado_municipio)

# --- NIVEL 3: mediana global (último recurso) ---
n_nivel3 = df['estrato'].isna().sum()
mediana_global = round(df.loc[reales, 'estrato'].median())
df['estrato'] = df['estrato'].fillna(mediana_global).astype(int)

total_imp = df['estrato_imputado'].sum()
print("═" * 58)
print("IMPUTACIÓN DE ESTRATO — CASCADA JERÁRQUICA")
print("═" * 58)
print(f"Estratos 0 (inválidos):        {n_estrato_0}")
print(f"Estratos faltantes (NaN):      {n_estrato_na}")
print(f"Total a imputar:               {total_imp}")
print("-" * 58)
print(f"  Nivel 1 — mediana del BARRIO:    {n_nivel1:>5}  ({n_nivel1/total_imp*100:>4.1f}%)")
print(f"  Nivel 2 — mediana del MUNICIPIO: {n_nivel2:>5}  ({n_nivel2/total_imp*100:>4.1f}%)")
print(f"  Nivel 3 — mediana GLOBAL:        {n_nivel3:>5}  ({n_nivel3/total_imp*100:>4.1f}%)")
print("-" * 58)
print(f"Barrios en la tabla de referencia: {len(tabla_barrio)}")
print(f"Nulos restantes en estrato:        {df['estrato'].isna().sum()}")
print()
print("Distribución final de estrato:")
print(df['estrato'].value_counts().sort_index().to_string())

# ═══════════════════════════════════════════════════════════════
# 2) MEDIO DE TRANSPORTE — Normalizar categorías
# ═══════════════════════════════════════════════════════════════
mapa_transporte = {
    # A pie
    'A Pie': 'A pie', 'Peatón': 'A pie',
    # Transporte público colectivo
    'Público Bus o Buseta': 'Transporte público', 'SITP o Bus': 'Transporte público',
    'Transmilenio': 'Transporte público', 'Masivo Integrado de Occidente': 'Transporte público',
    'Público Metro': 'Transporte público', 'Público Otro': 'Transporte público',
    # Vehículo propio
    'Moto': 'Moto', 'Particular Motocicleta': 'Moto',
    'Particular Automóvil': 'Automóvil', 'Carro Particular': 'Automóvil',
    'Bicicleta': 'Bicicleta', 'Particular Bicicleta': 'Bicicleta',
    # Taxi
    'Taxi': 'Taxi', 'Público Taxi': 'Taxi',
    # Otros
    'Ruta Escolar': 'Ruta escolar',
    'Otro': 'Otro', 'Particular Otro': 'Otro',
}

sin_mapear = set(df['medio_de_transporte'].dropna().unique()) - set(mapa_transporte)
if sin_mapear:
    print("\n⚠️ Categorías de transporte sin mapear (revisar):", sin_mapear)

df['medio_de_transporte'] = df['medio_de_transporte'].map(mapa_transporte)

print()
print("Medio de transporte normalizado (antes de imputar):")
print(df['medio_de_transporte'].value_counts(dropna=False).to_string())

══════════════════════════════════════════════════════════
IMPUTACIÓN DE ESTRATO — CASCADA JERÁRQUICA
══════════════════════════════════════════════════════════
Estratos 0 (inválidos):        46
Estratos faltantes (NaN):      1268
Total a imputar:               1314
----------------------------------------------------------
  Nivel 1 — mediana del BARRIO:      735  (55.9%)
  Nivel 2 — mediana del MUNICIPIO:   527  (40.1%)
  Nivel 3 — mediana GLOBAL:           52  ( 4.0%)
----------------------------------------------------------
Barrios en la tabla de referencia: 73
Nulos restantes en estrato:        0

Distribución final de estrato:
estrato
1     392
2    5844
3    6357
4     691
5     150
6      27

Medio de transporte normalizado (antes de imputar):
medio_de_transporte
Transporte público    4905
NaN                   3363
A pie                 3050
Moto                  1241
Automóvil              381
Bicicleta              272
Otro                   129
Taxi                    93
R

/tmp/ipykernel_872/2632846856.py:19: RuntimeWarning: Engine has switched to 'python' because numexpr does not support extension array dtypes. Please set your engine to python manually.
  .query(f'count >= {MIN_N}')['median'].round())  # redondeo: la mediana puede dar 2.5
/tmp/ipykernel_872/2632846856.py:30: RuntimeWarning: Engine has switched to 'python' because numexpr does not support extension array dtypes. Please set your engine to python manually.
  .query(f'count >= {MIN_N}')['median'].round())


##**Verificación: ¿la moda de transporte se sostiene al segmentar?**
Antes de imputar los nulos de transporte con la categoría más frecuente, comprobamos que esa moda
no sea un artefacto del promedio global. La duda es legítima: "A pie" representa casi un tercio de
los registros de Envigado, así que podría ser la opción dominante en algunos subgrupos.

La celda siguiente calcula la distribución de transporte dentro de cada municipio y cada zona.
Si "Transporte público" gana en todos los segmentos, la imputación es sólida, si algún grupo tuviera
otra moda, imputaríamos ese grupo con su propia moda (imputación condicional).

In [39]:
# Distribución de transporte por MUNICIPIO y por ZONA (solo registros con dato)

con_dato = df[df['medio_de_transporte'].notna()]

print("═" * 58)
print("MODA DE TRANSPORTE POR MUNICIPIO DE RESIDENCIA")
print("═" * 58)
for mun in df['municipio_direccion'].value_counts().head(5).index:
    sub = con_dato[con_dato['municipio_direccion'] == mun]
    if len(sub) < 30:
        continue
    dist = (sub['medio_de_transporte'].value_counts(normalize=True) * 100).round(1)
    print(f"\n{mun}  (n={len(sub)})  ->  moda: {dist.index[0]} ({dist.iloc[0]}%)")
    print(dist.head(3).to_string())

print("\n" + "═" * 58)
print("MODA DE TRANSPORTE POR ZONA")
print("═" * 58)
for z in con_dato['zona'].dropna().unique():
    sub = con_dato[con_dato['zona'] == z]
    if len(sub) < 30:
        continue
    dist = (sub['medio_de_transporte'].value_counts(normalize=True) * 100).round(1)
    print(f"\n{z}  (n={len(sub)})  ->  moda: {dist.index[0]} ({dist.iloc[0]}%)")
    print(dist.head(3).to_string())

══════════════════════════════════════════════════════════
MODA DE TRANSPORTE POR MUNICIPIO DE RESIDENCIA
══════════════════════════════════════════════════════════

Envigado  (n=8608)  ->  moda: Transporte público (47.8%)
medio_de_transporte
Transporte público    47.8
A pie                 32.4
Moto                  11.2

Medellín  (n=886)  ->  moda: Transporte público (47.9%)
medio_de_transporte
Transporte público    47.9
A pie                 26.3
Moto                  14.8

Itagui  (n=307)  ->  moda: Transporte público (59.6%)
medio_de_transporte
Transporte público    59.6
Moto                  21.2
Automóvil              6.8

Sabaneta  (n=140)  ->  moda: Transporte público (67.9%)
medio_de_transporte
Transporte público    67.9
Moto                  18.6
Automóvil              6.4

La Estrella  (n=43)  ->  moda: Transporte público (55.8%)
medio_de_transporte
Transporte público    55.8
Moto                  37.2
Automóvil              4.7

═══════════════════════════════════════════

#**Resultado de la verificación e imputación condicional**

**"Transporte público" es la moda en todos los segmentos** (Envigado 47.8%, Medellín 48.0%,
Itagüí 59.6%, Sabaneta 67.9%, zona urbana 46.8%, zona rural 63.2%). La preocupación por "A pie"
queda resuelta: aunque es relevante en Envigado (32.4%), nunca supera al transporte público.

**Decisión:** se imputan los nulos con la **moda del grupo al que pertenece cada estudiante**
(municipio de residencia, si el municipio no tiene muestra suficiente, se usa la moda global).
Como todos los grupos coinciden, el resultado equivale a imputar con "Transporte público", pero
la decisión queda respaldada por la verificación y no por un supuesto.

Se crea además la bandera `transporte_imputado`, por la misma razón que en el estrato: el 93.9%
de los cursos de extensión no registra transporte, así que el faltante es **sistemático** y esa
ausencia es información en sí misma.

In [40]:
MIN_N_TRANSPORTE = 30

# Bandera antes de imputar
df['transporte_imputado'] = df['medio_de_transporte'].isna().astype(int)
n_a_imputar = df['transporte_imputado'].sum()

# --- Moda condicional por municipio ---
modas_municipio = (df[df['medio_de_transporte'].notna()]
                   .groupby('municipio_direccion')['medio_de_transporte']
                   .agg(lambda s: s.mode().iloc[0] if len(s) >= MIN_N_TRANSPORTE else np.nan)
                   .dropna())

estimado = df['municipio_direccion'].map(modas_municipio)
n_por_grupo = (df['medio_de_transporte'].isna() & estimado.notna()).sum()
df['medio_de_transporte'] = df['medio_de_transporte'].fillna(estimado)

# --- Respaldo: moda global ---
moda_global = df.loc[df['transporte_imputado'] == 0, 'medio_de_transporte'].mode().iloc[0]
n_global = df['medio_de_transporte'].isna().sum()
df['medio_de_transporte'] = df['medio_de_transporte'].fillna(moda_global)

print("═" * 58)
print("IMPUTACIÓN DE TRANSPORTE — MODA CONDICIONAL")
print("═" * 58)
print(f"Registros a imputar:              {n_a_imputar}")
print(f"  Imputados por moda del MUNICIPIO: {n_por_grupo:>5}")
print(f"  Imputados por moda GLOBAL:        {n_global:>5}  (moda: {moda_global})")
print(f"Nulos restantes:                  {df['medio_de_transporte'].isna().sum()}")
print()
print("Distribución final de medio de transporte:")
print(df['medio_de_transporte'].value_counts().to_string())
print()
print("Bandera transporte_imputado:")
print(df['transporte_imputado'].value_counts().to_string())

══════════════════════════════════════════════════════════
IMPUTACIÓN DE TRANSPORTE — MODA CONDICIONAL
══════════════════════════════════════════════════════════
Registros a imputar:              3363
  Imputados por moda del MUNICIPIO:  3262
  Imputados por moda GLOBAL:          101  (moda: Transporte público)
Nulos restantes:                  0

Distribución final de medio de transporte:
medio_de_transporte
Transporte público    8268
A pie                 3050
Moto                  1241
Automóvil              381
Bicicleta              272
Otro                   129
Taxi                    93
Ruta escolar            27

Bandera transporte_imputado:
transporte_imputado
0    10098
1     3363


In [41]:
df.head(1000)

,genero,municipio_de_nacimiento,pais,municipio_direccion,barrio_direccion,estrato,estado_civil,eps,grupo_sisben,zona,...,jornada,programa,periodo,nivel,edad,periodo_era_mt,anio_periodo,semestre_periodo,estrato_imputado,transporte_imputado
0,Femenino,Medellín,Colombia,Envigado,Las Flores,3,Soltero(a),EPS Sura,NaN,Urbana,...,MEDIA TÉCNICA,MEDIA TÉCNICA TECNICO LABORAL AUXILIAR ADMINIS...,2018-2,SEMESTRE 2,20,0,2018,2,0,0
1,Femenino,Envigado,Colombia,Envigado,San José,2,Soltero(a),EPS Sura,NaN,Urbana,...,NOCHE,10T-TÉCNICO LABORAL EN AUXILIAR ADMINISTRATIVO...,2016-2,SEMESTRE 2,29,0,2016,2,0,0
2,Masculino,Bogotá,Colombia,La Mesa,NaN,3,Soltero(a),NaN,NaN,Rural,...,NOCHE,09T-TÉCNICO LABORAL EN AUXILIAR ADMINISTRATIVO,2021-2,SEMESTRE 1,19,0,2021,2,0,0
3,Masculino,Envigado,Colombia,Envigado,San José,2,Soltero(a),SISBEN,B3,Urbana,...,DIURNA,01T-TÉCNICO LABORAL AUXILIAR DE TALENTO HUMANO,2021-1,SEMESTRE 2,21,0,2021,1,0,0
4,Masculino,Medellín,Colombia,Envigado,El Salado,2,Soltero(a),Nueva Promotora de Salud - Nueva EPS,N0,Urbana,...,NOCHE,27T-TÉCNICO LABORAL EN MERCADEO Y VENTAS (Estu...,2020-1,SEMESTRE 1,20,0,2020,1,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,Femenino,Envigado,Colombia,Envigado,Las Orquídeas,2,Soltero(a),SISBEN,NaN,Urbana,...,MEDIA TÉCNICA,MEDIA TECNICA TECNICO LABORAL EN ENTRENAMIENTO...,2021-1,ONCE,17,1,2021,1,0,0
996,Femenino,Itagui,Colombia,Envigado,La Magnolia,3,Soltero(a),EPS Sura,NaN,Urbana,...,NOCHE,03T-TÉCNICO LABORAL AUXILIAR EN LOGÍSTICA,2020-1,SEMESTRE 1,19,0,2020,1,0,0
997,Femenino,San Vicente,Colombia,Envigado,Barrio Mesa,3,Soltero(a),NaN,NaN,Urbana,...,MEDIA TÉCNICA,PROYECTOS FORMATIVOS MEDIA TECNICA,2018-2,ONCE,20,1,2018,2,0,0
998,Masculino,Envigado,Colombia,Envigado,El Salado,2,Soltero(a),EPS Sura,NaN,Urbana - rural,...,NOCHE,17T-TÉCNICO LABORAL EN ENTRENAMIENTO DEPORTIVO,2016-1,SEMESTRE 2,26,0,2016,1,0,0


## Columnas derivadas de la fecha

Extraemos componentes de la fecha que pueden ser útiles
para el EDA y el modelado.

In [42]:
df['anio_matricula'] = df['fecha_de_matricula'].dt.year
df['mes_matricula']  = df['fecha_de_matricula'].dt.month
df[['fecha_de_matricula', 'anio_matricula', 'mes_matricula']].head()

,fecha_de_matricula,anio_matricula,mes_matricula
0,2018-02-05 10:00:00,2018.0,2.0
1,2016-01-19 10:00:00,2016.0,1.0
2,2021-02-08 10:00:00,2021.0,2.0
3,2020-01-17 10:00:00,2020.0,1.0
4,2020-01-27 10:00:00,2020.0,1.0


In [43]:
df

,genero,municipio_de_nacimiento,pais,municipio_direccion,barrio_direccion,estrato,estado_civil,eps,grupo_sisben,zona,...,periodo,nivel,edad,periodo_era_mt,anio_periodo,semestre_periodo,estrato_imputado,transporte_imputado,anio_matricula,mes_matricula
0,Femenino,Medellín,Colombia,Envigado,Las Flores,3,Soltero(a),EPS Sura,NaN,Urbana,...,2018-2,SEMESTRE 2,20,0,2018,2,0,0,2018.0,2.0
1,Femenino,Envigado,Colombia,Envigado,San José,2,Soltero(a),EPS Sura,NaN,Urbana,...,2016-2,SEMESTRE 2,29,0,2016,2,0,0,2016.0,1.0
2,Masculino,Bogotá,Colombia,La Mesa,NaN,3,Soltero(a),NaN,NaN,Rural,...,2021-2,SEMESTRE 1,19,0,2021,2,0,0,2021.0,2.0
3,Masculino,Envigado,Colombia,Envigado,San José,2,Soltero(a),SISBEN,B3,Urbana,...,2021-1,SEMESTRE 2,21,0,2021,1,0,0,2020.0,1.0
4,Masculino,Medellín,Colombia,Envigado,El Salado,2,Soltero(a),Nueva Promotora de Salud - Nueva EPS,N0,Urbana,...,2020-1,SEMESTRE 1,20,0,2020,1,0,0,2020.0,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
13456,Masculino,Envigado,Colombia,Envigado,Barrio Mesa,3,Soltero(a),21) EPS Sura,D21,Urbana,...,2026-MT,ONCE,16,1,2026,<NA>,0,0,NaN,NaN
13457,Masculino,NaN,Colombia,Envigado,La Magnolia,3,NaN,COOSALUD EPS,B3,Urbana,...,2026-MT,ONCE,16,1,2026,<NA>,0,0,NaN,NaN
13458,Masculino,NaN,Colombia,Envigado,San José,2,NaN,NaN,NaN,NaN,...,2026-MT,ONCE,16,1,2026,<NA>,1,1,NaN,NaN
13459,Femenino,Envigado,Colombia,Envigado,La Mina,3,Soltero(a),21) EPS Sura,N0,Urbana,...,2026-MT,ONCE,16,1,2026,<NA>,0,0,NaN,NaN


### Definición adoptada (criterio conservador + regla temporal)
> Se considera **deserción (`1`)** cuando hay **abandono explícito** de los estudios, o
> cuando un estudiante en estado *temporal* (inactivo/aplazado) lleva **más de 2 años** sin
> volver a matricularse
> si tras cierto tiempo no hay reingreso, se considera abandono. Se considera **no deserción
> (`0`)** cuando el estudiante continúa, culminó, o lleva **menos de 2 años** en estado
> temporal (aún podría reingresar).

| Estado original | Clasificación | Justificación |
|---|---|---|
| Graduado, Egresado | **0** | Culminó. *Egresado y Graduado se tratan como equivalentes.* |
| Activo | **0** | Continúa matriculado. |
| Cancelado, Desertor, Retiro - Aplazados, Sancionado | **1** | Abandono explícito / no continuación. |
| **Inactivo, Aplazamiento Semestre, Aplazamiento Actividades** | **regla de 2 años** | `1` si la última matrícula fue hace **más de 2 años**; `0` si fue hace **menos de 2 años**. |
| Sin estado | **excluido (NaN)** | Sin información. |


In [44]:
# Valores reales de la variable de estado ANTES de mapear
df['estado'].value_counts(dropna=False)

,count
estado,
Inactivo,4838
Graduado,3831
Activo,2808
Cancelado,759
Sancionado,602
Aplazamiento Semestre,285
Sin estado,243
Desertor,79
Aplazamiento Actividades,8


In [45]:
mapa_desercion = {
    # --- No deserción (0): continúa o culminó ---
    'Graduado': 0,
    'Egresado': 0,
    'Activo':   0,
    # --- Deserción (1): abandono explícito ---
    'Cancelado':          1,
    'Desertor':           1,
    'Retiro - Aplazados': 1,
    'Sancionado':         1,
    # --- Sin información: se excluye ---
    'Sin estado': np.nan,
}

df['desercion'] = df['estado'].map(mapa_desercion)

In [46]:
df

,genero,municipio_de_nacimiento,pais,municipio_direccion,barrio_direccion,estrato,estado_civil,eps,grupo_sisben,zona,...,nivel,edad,periodo_era_mt,anio_periodo,semestre_periodo,estrato_imputado,transporte_imputado,anio_matricula,mes_matricula,desercion
0,Femenino,Medellín,Colombia,Envigado,Las Flores,3,Soltero(a),EPS Sura,NaN,Urbana,...,SEMESTRE 2,20,0,2018,2,0,0,2018.0,2.0,0.0
1,Femenino,Envigado,Colombia,Envigado,San José,2,Soltero(a),EPS Sura,NaN,Urbana,...,SEMESTRE 2,29,0,2016,2,0,0,2016.0,1.0,0.0
2,Masculino,Bogotá,Colombia,La Mesa,NaN,3,Soltero(a),NaN,NaN,Rural,...,SEMESTRE 1,19,0,2021,2,0,0,2021.0,2.0,1.0
3,Masculino,Envigado,Colombia,Envigado,San José,2,Soltero(a),SISBEN,B3,Urbana,...,SEMESTRE 2,21,0,2021,1,0,0,2020.0,1.0,0.0
4,Masculino,Medellín,Colombia,Envigado,El Salado,2,Soltero(a),Nueva Promotora de Salud - Nueva EPS,N0,Urbana,...,SEMESTRE 1,20,0,2020,1,0,0,2020.0,1.0,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
13456,Masculino,Envigado,Colombia,Envigado,Barrio Mesa,3,Soltero(a),21) EPS Sura,D21,Urbana,...,ONCE,16,1,2026,<NA>,0,0,NaN,NaN,0.0
13457,Masculino,NaN,Colombia,Envigado,La Magnolia,3,NaN,COOSALUD EPS,B3,Urbana,...,ONCE,16,1,2026,<NA>,0,0,NaN,NaN,0.0
13458,Masculino,NaN,Colombia,Envigado,San José,2,NaN,NaN,NaN,NaN,...,ONCE,16,1,2026,<NA>,1,1,NaN,NaN,0.0
13459,Femenino,Envigado,Colombia,Envigado,La Mina,3,Soltero(a),21) EPS Sura,N0,Urbana,...,ONCE,16,1,2026,<NA>,0,0,NaN,NaN,0.0


### Regla de continuidad para los estados ambiguos
Para `Inactivo`, `Aplazamiento Semestre` y `Aplazamiento Actividades` aplicamos la regla de
los 2 años: si la última matrícula fue hace **más de 2 años** se considera deserción (`1`);
si fue hace **menos de 2 años**, no deserción (`0`). Si no hay fecha válida, se excluye.

In [47]:
FECHA_CORTE = pd.Timestamp('2022-06-29').normalize()
LIMITE_2A   = FECHA_CORTE - pd.DateOffset(years=2)

estados_temporales = ['Inactivo', 'Aplazamiento Semestre', 'Aplazamiento Actividades']
mask_temporal = df['estado'].isin(estados_temporales)

# Más de 2 años sin volver a matricularse -> deserción (1)
df.loc[mask_temporal & (df['fecha_de_matricula'] <  LIMITE_2A), 'desercion'] = 1
# Menos de 2 años -> aún podría reingresar -> no deserción (0)
df.loc[mask_temporal & (df['fecha_de_matricula'] >= LIMITE_2A), 'desercion'] = 0


print('Fecha de corte:', FECHA_CORTE.date(), '| Umbral 2 años:', LIMITE_2A.date())
print()
for est in estados_temporales:
    sub = df[df['estado'] == est]
    a_1 = (sub['fecha_de_matricula'] <  LIMITE_2A).sum()
    a_0 = (sub['fecha_de_matricula'] >= LIMITE_2A).sum()
    nat = sub['fecha_de_matricula'].isna().sum()
    print(f'{est:<26} +2años->1: {a_1:>5} | <2años->0: {a_0:>4} | sin fecha: {nat}')

# Verificar que no quedó ningún estado sin clasificar por error de tipeo
clasificados = set(mapa_desercion) | set(estados_temporales)
sin_clasificar = set(df['estado'].dropna().unique()) - clasificados
print('\nEstados sin clasificar (debe ser vacío):', sin_clasificar)

Fecha de corte: 2022-06-29 | Umbral 2 años: 2020-06-29

Inactivo                   +2años->1:  3000 | <2años->0: 1715 | sin fecha: 123
Aplazamiento Semestre      +2años->1:   171 | <2años->0:  114 | sin fecha: 0
Aplazamiento Actividades   +2años->1:     3 | <2años->0:    5 | sin fecha: 0

Estados sin clasificar (debe ser vacío): set()


### Resumen del target resultante
Definición elegida: cuántas filas quedan etiquetadas,
cuántas se excluyen y el balance de clases (relevante para el modelado posterior).

In [48]:
n_total = len(df)
n_etiquetadas = df['desercion'].notna().sum()
n_excluidas = df['desercion'].isna().sum()

print(f"Filas totales:      {n_total}")
print(f"Etiquetadas:        {n_etiquetadas}")
print(f"Excluidas (NaN):    {n_excluidas}")
print()
print("Distribución del target (solo etiquetadas):")
print(df['desercion'].value_counts(dropna=False))
print()
tasa = df['desercion'].mean()
print(f"Tasa de deserción:  {tasa:.1%}")
print()

Filas totales:      13461
Etiquetadas:        13095
Excluidas (NaN):    366

Distribución del target (solo etiquetadas):
desercion
0.0    8480
1.0    4615
NaN     366
Name: count, dtype: int64

Tasa de deserción:  35.2%



## Revisión de duplicados

El dataset son *matrículas*, no estudiantes únicos. Se revisará filas
completamente duplicadas.

In [49]:
n_dup = df.duplicated().sum()
print("Filas duplicadas exactas:", n_dup)
df = df.drop_duplicates().reset_index(drop=True)
print("Shape tras eliminar duplicados:", df.shape)

Filas duplicadas exactas: 65
Shape tras eliminar duplicados: (13396, 29)


Anteriormente se detectaron 64 filas duplicadas, pero al ejecutar el código "df = df.drop_duplicates().reset_index(drop=True)" se eliminan las filas duplicadas.

In [50]:
filas_duplicadas = df[df.duplicated(keep=False)].sort_values(by=df.columns.tolist())
print(f"Total de filas duplicadas: {len(filas_duplicadas)}")
print()
filas_duplicadas

Total de filas duplicadas: 0



,genero,municipio_de_nacimiento,pais,municipio_direccion,barrio_direccion,estrato,estado_civil,eps,grupo_sisben,zona,...,nivel,edad,periodo_era_mt,anio_periodo,semestre_periodo,estrato_imputado,transporte_imputado,anio_matricula,mes_matricula,desercion


## 7. Columnas con exceso de nulos

> **Retroalimentación:** algunas variables se caerán por su alto % de nulos; el dataset se
> reducirá en columnas. Aquí identificamos y descartamos las que están **casi vacías**
> (la decisión final sobre el resto se toma en el EDA con pruebas estadísticas).

In [51]:
nulos_pct = (df.isnull().mean() * 100).round(1).sort_values(ascending=False)
print("% de nulos por columna:")
print(nulos_pct.head(8))

# Descartar columnas con >90% de nulos (aportan casi nula información)
umbral = 70
a_descartar = nulos_pct[nulos_pct > umbral].index.tolist()
print(f"\nColumnas descartadas (>{umbral}% nulos): {a_descartar}")
df = df.drop(columns=a_descartar)
print("Shape tras descartar:", df.shape)

% de nulos por columna:
multiculturalidad     93.8
grupo_sisben          76.8
nivel_de_formacion    24.9
eps                   22.6
ocupacion             14.2
estado_civil          10.7
barrio_direccion       9.8
zona                   9.5
dtype: float64

Columnas descartadas (>70% nulos): ['multiculturalidad', 'grupo_sisben']
Shape tras descartar: (13396, 27)


## Guardar los datos limpios

Guardar el dataset limpio, estandarizado y con la variable objetivo.

In [52]:
from getpass import getpass

ruta_salida = processed_dir + 'matriculas_limpias.csv'
df.to_csv(ruta_salida, index=False)
print("Guardado en:", ruta_salida)
print("Dimensiones finales:", df.shape)

!git config --global user.email "betancourt1659@gmail.com"
!git config --global user.name  "LuisBetancourt11"

usuario = "LuisBetancourt11"
repo    = "Proyecto_Ciencia_Datos"
token   = getpass("Access Token: ")

!cd {repo} && git add data/processed/matriculas_limpias.csv && \
  git commit -m "Limpieza corregida)" && \
  git remote set-url origin https://{usuario}:{token}@github.com/{usuario}/{repo}.git && \
  git push origin main

Guardado en: Proyecto_Ciencia_Datos/data/processed/matriculas_limpias.csv
Dimensiones finales: (13396, 27)
Access Token: ··········
[main 473de17] Limpieza corregida)
 1 file changed, 13397 insertions(+), 13397 deletions(-)
Enumerating objects: 9, done.
Counting objects: 100% (9/9), done.
Delta compression using up to 2 threads
Compressing objects: 100% (5/5), done.
Writing objects: 100% (5/5), 263.96 KiB | 1.53 MiB/s, done.
Total 5 (delta 2), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (2/2), completed with 2 local objects.
To https://github.com/LuisBetancourt11/Proyecto_Ciencia_Datos.git
   64c07ee..473de17  main -> main
